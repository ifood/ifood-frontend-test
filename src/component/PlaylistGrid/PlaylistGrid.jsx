import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import Playlist from '../Playlist';
import Info from '../Info';
import Grid from '../Grid';

function itemRenderer(playlist) {
  return <Playlist key={playlist?.id} playlist={playlist} />;
}

function PlaylistGrid({ title, items }) {
  const intl = useIntl();

  const gridTitle =
    title != null ? title : intl.formatMessage({ id: 'featured.playlists' });

  if (items == null || items.length === 0) {
    return (
      <Info
        icon="search"
        title={intl.formatMessage({ id: 'no.playlist' })}
        info={intl.formatMessage({ id: 'choose.filters.message' })}
      />
    );
  }

  return <Grid title={gridTitle} items={items} itemRenderer={itemRenderer} />;
}

PlaylistGrid.propTypes = {
  title: PropTypes.string,
  items: PropTypes.array,
};

export default React.memo(PlaylistGrid);
