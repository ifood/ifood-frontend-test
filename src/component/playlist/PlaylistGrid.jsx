/* eslint-disable react-hooks/exhaustive-deps, camelcase */
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
  Accordion,
  IconButton,
  TextField,
  Typography,
} from '@material-ui/core';
import { FilterList } from '@material-ui/icons';
import GridList from '../gridlist/GridList';
import PlaylistCard from './PlaylistCard';
import { CenteredFlex, LayoutContent } from '../containers/Containers';
import FilterForm from '../filter/FilterForm';
import PlaylistError from './PlaylistError';

const Content = styled(LayoutContent)`
  margin-top: 150px;
`;
const FilterContainer = styled.div`
  display: flex;
  padding-bottom: 16px;
`;
const AccordionContainer = styled.div`
  padding-bottom: 16px;
`;

function itemRenderer(element) {
  return <PlaylistCard playlist={element} />;
}

function PlaylistGrid(props) {
  const {
    data,
    errorMessage,
    filterData,
    filters,
    onChangeFilter,
    title,
  } = props;
  const [filterText, setFilterText] = useState('');
  const [gridData, setGridData] = useState([]);
  const [expandedFilter, setExpandedFilter] = useState(false);

  useEffect(() => {
    setGridData([...data]);
  }, [data]);

  const handleChangeTextFilter = useCallback((element) => {
    setFilterText(element?.target?.value);
  }, []);

  const handleExpandFilter = useCallback(() => {
    setExpandedFilter(!expandedFilter);
  }, [expandedFilter]);

  useEffect(() => {
    const filterValues = data.filter(
      (element) =>
        element.name.toUpperCase().includes(filterText.toUpperCase()) ||
        element.description.toUpperCase().includes(filterText.toUpperCase()),
    );

    setGridData(filterValues);
  }, [filterText]);

  const renderGrid = useMemo(() => {
    if (errorMessage != null) {
      return <PlaylistError error={errorMessage} />;
    }

    if (gridData.length > 0) {
      return (
        <>
          <Typography color="textPrimary" variant="h5" gutterBottom>
            {title}
          </Typography>
          <GridList
            data={gridData}
            itemRenderer={itemRenderer}
            filters={filters}
          />
        </>
      );
    }

    return (
      <CenteredFlex>
        <Typography color="textPrimary" variant="h5" gutterBottom>
          Your search result did not return any playlist
        </Typography>
      </CenteredFlex>
    );
  }, [title, gridData, itemRenderer, filters, errorMessage]);

  return (
    <Content>
      <AccordionContainer>
        <Accordion expanded={expandedFilter}>
          <FilterContainer>
            <TextField
              id="local-filter"
              label="Search playlist"
              variant="filled"
              onChange={handleChangeTextFilter}
              fullWidth
            />
            <IconButton color="primary" onClick={handleExpandFilter}>
              <FilterList />
            </IconButton>
          </FilterContainer>
          <FilterForm
            filters={filters}
            onChange={onChangeFilter}
            filterData={filterData}
          />
        </Accordion>
      </AccordionContainer>
      {renderGrid}
    </Content>
  );
}

PlaylistGrid.defaultProps = {
  data: [],
};

PlaylistGrid.propTypes = {
  data: PropTypes.array,
  title: PropTypes.string,
  errorMessage: PropTypes.string,
  filters: PropTypes.array,
  filterData: PropTypes.object,
  onChangeFilter: PropTypes.func,
};

export default PlaylistGrid;
