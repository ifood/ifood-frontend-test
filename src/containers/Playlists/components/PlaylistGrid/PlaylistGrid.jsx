
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import GridList from '@material-ui/core/GridList';
import GridListTile from '@material-ui/core/GridListTile';
import GridListTileBar from '@material-ui/core/GridListTileBar';

const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  margin-top: 25px;
`;

const Link = styled.a`
  text-decoration: none;
  display: block;
`;

const StyledGridListTileBar = styled(GridListTileBar)`
  max-width: 300px;
`;

class PlaylistGrid extends Component {
  constructor(props) {
    super(props);
    this.state = {
      columns: 0,
    };
  }

  componentDidMount() {
    this.evaluateColumnsNumber();
    window.addEventListener('resize', this.evaluateColumnsNumber);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.evaluateColumnsNumber);
  }

  evaluateColumnsNumber = () => {
    this.setState({
      columns: Math.floor(window.innerWidth / 300) || 1,
    });
  }

  render() {
    const { playlists } = this.props;
    const { columns } = this.state;
    const cols = columns <= playlists.length ? columns : playlists.length;

    return (
      <Container>
        <GridList cellHeight={300} cols={cols}>
          {playlists.map(playlist => (
            <GridListTile key={playlist.images[0].url} cols={1}>
              <Link href={playlist.external_urls.spotify} target="_blank">
                <img src={playlist.images[0].url} alt="playlist" />
                <StyledGridListTileBar
                  title={playlist.name}
                  subtitle={playlist.owner.display_name}
                />
              </Link>
            </GridListTile>
          ))}
        </GridList>
      </Container>
    );
  }
}

PlaylistGrid.propTypes = {
  playlists: PropTypes.array,
};

PlaylistGrid.defaultProps = {
  playlists: [],
};

export default PlaylistGrid;
