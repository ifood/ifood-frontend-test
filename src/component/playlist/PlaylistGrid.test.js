/* eslint-disable no-undef */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import PlaylistGrid from './PlaylistGrid';

describe('<PlaylistGrid />', () => {
  it('snapshot', async () => {
    const wrapper = mount(<PlaylistGrid />);
    jestExpect(toJson(wrapper)).toMatchSnapshot();
  });
});
