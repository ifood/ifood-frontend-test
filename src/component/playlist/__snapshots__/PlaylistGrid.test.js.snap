// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaylistError /> snapshot 1`] = `
.c0 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  background-color: #0f0f0f;
  height: 100%;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
}

.c1 {
  margin-top: 80px;
}

<PlaylistError>
  <Styled(styled.div)>
    <div
      className="c0 c1"
    >
      <styled.div>
        <div
          className="c2"
        >
          <ForwardRef
            color="primary"
          >
            <WithStyles(ForwardRef(SvgIcon))
              color="primary"
            >
              <ForwardRef(SvgIcon)
                classes={
                  Object {
                    "colorAction": "MuiSvgIcon-colorAction",
                    "colorDisabled": "MuiSvgIcon-colorDisabled",
                    "colorError": "MuiSvgIcon-colorError",
                    "colorPrimary": "MuiSvgIcon-colorPrimary",
                    "colorSecondary": "MuiSvgIcon-colorSecondary",
                    "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                    "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                    "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                    "root": "MuiSvgIcon-root",
                  }
                }
                color="primary"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
                  />
                </svg>
              </ForwardRef(SvgIcon)>
            </WithStyles(ForwardRef(SvgIcon))>
          </ForwardRef>
          <WithStyles(ForwardRef(Typography))
            color="textPrimary"
          >
            <ForwardRef(Typography)
              classes={
                Object {
                  "alignCenter": "MuiTypography-alignCenter",
                  "alignJustify": "MuiTypography-alignJustify",
                  "alignLeft": "MuiTypography-alignLeft",
                  "alignRight": "MuiTypography-alignRight",
                  "body1": "MuiTypography-body1",
                  "body2": "MuiTypography-body2",
                  "button": "MuiTypography-button",
                  "caption": "MuiTypography-caption",
                  "colorError": "MuiTypography-colorError",
                  "colorInherit": "MuiTypography-colorInherit",
                  "colorPrimary": "MuiTypography-colorPrimary",
                  "colorSecondary": "MuiTypography-colorSecondary",
                  "colorTextPrimary": "MuiTypography-colorTextPrimary",
                  "colorTextSecondary": "MuiTypography-colorTextSecondary",
                  "displayBlock": "MuiTypography-displayBlock",
                  "displayInline": "MuiTypography-displayInline",
                  "gutterBottom": "MuiTypography-gutterBottom",
                  "h1": "MuiTypography-h1",
                  "h2": "MuiTypography-h2",
                  "h3": "MuiTypography-h3",
                  "h4": "MuiTypography-h4",
                  "h5": "MuiTypography-h5",
                  "h6": "MuiTypography-h6",
                  "noWrap": "MuiTypography-noWrap",
                  "overline": "MuiTypography-overline",
                  "paragraph": "MuiTypography-paragraph",
                  "root": "MuiTypography-root",
                  "srOnly": "MuiTypography-srOnly",
                  "subtitle1": "MuiTypography-subtitle1",
                  "subtitle2": "MuiTypography-subtitle2",
                }
              }
              color="textPrimary"
            >
              <p
                className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextPrimary"
              />
            </ForwardRef(Typography)>
          </WithStyles(ForwardRef(Typography))>
        </div>
      </styled.div>
    </div>
  </Styled(styled.div)>
</PlaylistError>
`;
